Grails 5.2 app with Micronaut beans will not run with embedded Tomcat

### Expected Behavior

The application should start

### Actual Behaviour

The application fails with:

```
java.nio.file.ProviderNotFoundException: null
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:603)
        at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:678)
        at java.base/java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:722)
        at io.micronaut.core.io.service.SoftServiceLoader$DefaultServiceCollector.collect(SoftServiceLoader.java:536)
        at io.micronaut.core.io.service.SoftServiceLoader.collectDynamicServices(SoftServiceLoader.java:199)
        at io.micronaut.core.io.service.SoftServiceLoader.collectAll(SoftServiceLoader.java:175)
        at io.micronaut.context.DefaultBeanContext.resolveBeanDefinitionReferences(DefaultBeanContext.java:1732)
        at io.micronaut.context.DefaultApplicationContext.isBootstrapPropertySourceLocatorPresent(DefaultApplicationContext.java:153)
        at io.micronaut.context.DefaultApplicationContext.isBootstrapEnabled(DefaultApplicationContext.java:149)
        at io.micronaut.context.DefaultApplicationContext.createEnvironment(DefaultApplicationContext.java:137)
        at io.micronaut.context.DefaultApplicationContext.getEnvironment(DefaultApplicationContext.java:181)
        at grails.boot.GrailsApp.createApplicationContext(GrailsApp.groovy:181)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:305)
        at grails.boot.GrailsApp.run(GrailsApp.groovy:99)
        at grails.boot.GrailsApp.run(GrailsApp.groovy:485)
        at grails.boot.GrailsApp.run(GrailsApp.groovy:472)
        at gr52.mn.bug.Application.main(Application.groovy:11)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:108)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:58)
        at org.springframework.boot.loader.WarLauncher.main(WarLauncher.java:59)
Caused by: java.nio.file.ProviderNotFoundException: Provider not found
        at java.base/java.nio.file.FileSystems.newFileSystem(FileSystems.java:423)
        at io.micronaut.core.io.IOUtils.loadNestedJarUri(IOUtils.java:153)
        at io.micronaut.core.io.IOUtils.eachFile(IOUtils.java:99)
        at io.micronaut.core.io.service.SoftServiceLoader.computeServiceTypeNames(SoftServiceLoader.java:277)
        at io.micronaut.core.io.service.SoftServiceLoader.access$1100(SoftServiceLoader.java:68)
        at io.micronaut.core.io.service.SoftServiceLoader$MicronautMetaServicesLoader.compute(SoftServiceLoader.java:565)
        at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:189)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

### Steps To Reproduce

1. Create new Grails 5.2.1 application
2. `./gradlew bootWar`
3. `java -jar build/libs/gr52-mn-bug-0.1.war`
4. Application Starts
5. Add a Micronaut bean
6. `./gradlew clean bootWar`
7. `java -jar build/libs/gr52-mn-bug-0.1.war`
8. Observe the exception above

### Environment Information

- Debian 11
- Grails 5.2.1
- openjdk version "11.0.12" 2021-07-20

### Example Application

https://github.com/JasonTypesCodes/gr52-mn-bug

### Version

5.2.1

--- a/core/src/main/java/io/micronaut/core/io/IOUtils.java
+++ b/core/src/main/java/io/micronaut/core/io/IOUtils.java
@@ -136,6 +136,11 @@ public class IOUtils {
             return Paths.get(URI.create(jarUri));
         }
         Path jarPath = loadNestedJarUri(toClose, jarUri.substring(0, sep));
+        if (Files.isDirectory(jarPath)) {
+            // spring boot creates weird jar URLs, like 'jar:file:/xyz.jar!/BOOT-INF/classes!/abc'
+            // This check makes our class loading resilient to that
+            return jarPath;
+        }
         FileSystem zipfs;
         try {
             // can't use newFileSystem(Path) here (without CL) because it doesn't exist on java 8

--- a/core/src/test/groovy/io/micronaut/core/io/IOUtilsSpec.groovy
+++ b/core/src/test/groovy/io/micronaut/core/io/IOUtilsSpec.groovy
@@ -1,5 +1,6 @@
 package io.micronaut.core.io
 
+import spock.lang.Issue
 import spock.lang.Specification
 
 import java.nio.charset.StandardCharsets
@@ -61,6 +62,32 @@ class IOUtilsSpec extends Specification {
         })
         then:
         visitedInner == ['baz.txt']
+        textInner == ['bla']
+
+        cleanup:
+        Files.deleteIfExists(zipPath)
+    }
+
+    @Issue('https://github.com/grails/grails-core/issues/12625/')
+    def 'dir inside jar'() {
+        given:
+        Path zipPath = Files.createTempFile("micronaut-ioutils-spec", ".zip")
+        try (ZipOutputStream outer = new ZipOutputStream(Files.newOutputStream(zipPath))) {
+            outer.putNextEntry(new ZipEntry("foo/bar/baz/test.txt"))
+            outer.write("bla".getBytes(StandardCharsets.UTF_8))
+            outer.closeEntry()
+        }
+
+        def visitedInner = []
+        def textInner = []
+
+        when:
+        IOUtils.eachFile(URI.create('jar:' + zipPath.toUri() + '!/foo/bar!/xyz'), 'baz', entry -> {
+            visitedInner.add(entry.getFileName().toString())
+            textInner = Files.readAllLines(entry)
+        })
+        then:
+        visitedInner == ['test.txt']
         textInner == ['bla']
 
         cleanup:

